src/symbol/get_anchors.js

```diff
@@ -59,6 +59,16 @@ function getCenterAnchor(line: Array<Point>,
 
             const anchor = new Anchor(x, y, b.angleTo(a), i);
             anchor._round();
+            // If the anchor point is very close to a or b, its x or y coordinate may become exactly the same as a or b
+            // after rounding. In this case, we will have to update the anchor point to be equal to a or b. Otherwise, the
+            // anchor point will not be on the line, further leading to angle calculation error when projecting symbols.
+            if (t > 0.99 && (anchor.x === b.x || anchor.y === b.y)) {
```

Comment:
Are the 0.01/0.99 thresholds general? Would dependence on the length of the line either too often apply this for long lines or not apply this for short lines?

To check if it the anchor is within a pixel of the endpoint, I tried out:

```js
if (t * segmentDistance > segmentDistance - 1 && (anchor.x === b.x || anchor.y === b.y)) {
  ...
} else if (t * segmentDistance < 1 && (anchor.x === a.x || anchor.y === a.y)) {
  ...
} 
```

and it seemed to work, but I'm not certain this is actually better or more general.

Reply:
I am thinking `t * segmentDistance > segmentDistance - 1` maybe two strong, the same as for `0.01/009`. I took them to only indicate which point will the anchor be closed to. So I wander should we just use 0.5 as the loose comparison, the further anchor point comparison will anyway do the enhanced check. wdyt?

Reply:
Ah, I see! My main concern was that the 0.01/0.99 threshold might lead to false positives/negatives in particular, I think, when the two points are close together. Just to be certain, would 0.5 make it possible for a perfectly horizontal line to pass the check incorrectly? i.e. `anchor.y === b.y` when the points are not actually close together?

Reply:
I think almost all of my remaining confusion comes down to this single condition:

```js
anchor.x === b.x || anchor.y === b.y
```

Would a perfectly horizontal line (anchor.x = b.x) pass this condition when the two points are not actually close together?

Reply:
nice point! The same for peectly vertical line. Since the anchor is calculated as `(vectorB - vectorA) * t`, which is `[b.x-a.x, b.y - a.y] * t` => `[dx, dy] * t`
so can we use following check, which is inspried from your solution:
```
if (dx !== 0 && dy !== 0) // exclude perfectly vertical/horizontal case {
    if ((dx * t < 1 || dy * t < 1) && (anchor.x === a.x || anchor.y ==== a.y) ) { 
        anchor.x = a.x;
        anchor.y = a.y;
    } else if ((dx * t > dx - 1 || dy * t > dy - 1) && (anchor.x === b.x || anchor.y ==== b.y) ) { 
        anchor.x = b.x;
        anchor.y = b.y;
    }
}
```

Reply:
@rreusser, thanks to @mpulkki-mapbox 's suggestion, I think the best option would be just removing the rounding process, and keep the resolution of the anchor as it is.

Reply:
That seems like a nice solution! I don't think there are generally thousands of characters, so it's hard to imagine that upgrading a single number from two bytes to four would have a detrimental effect on performance. Sorry to drag my heels on this one, but it seemed tricky enough that I wanted to make sure the motivation and behavior was _really_ clear. This seems like a reasonable solution. 👍